crates/iota-indexer/src/optimistic_indexing.rs

```diff
@@ -0,0 +1,393 @@
+        let backoff = backoff::ExponentialBackoff {
+            max_elapsed_time: Some(Duration::from_secs(30)),
+            ..Default::default()
+        };
+
+        backoff::future::retry(backoff, async || {
+            let tx_block_response = self
+                .indexer_reader
+                .multi_get_transaction_block_response_in_blocking_task(
+                    vec![tx_digest],
+                    options.clone().unwrap_or_default(),
+                )
+                .await
+                .map_err(|e| backoff::Error::Transient {
+                    err: e,
+                    retry_after: None,
+                })?
+                .pop();
+
+            match tx_block_response {
+                Some(tx_block_response) => Ok(tx_block_response),
+                None => Err(backoff::Error::Transient {
+                    err: IndexerError::PostgresRead("Transaction not present in DB".to_string()),
+                    retry_after: None,
+                }),
+            }
+        })
+        .await
+    }
+
+    async fn index_transaction(
+        &self,
+        full_tx_data: &CheckpointTransaction,
+    ) -> Result<(), IndexerError> {
+        let assigned_global_order = self
+            .assign_optimistic_tx_global_order(full_tx_data.transaction.digest())
+            .await?;
+
+        let Some(assigned_global_order) = assigned_global_order else {
+            // Global order was assigned earlier by other indexing process, we avoid double
+            // or concurrent indexing and return
+            return Ok(());
+        };
+
+        let extractor = TransactionExtractor::new(
+            full_tx_data,
+            assigned_global_order
+                .optimistic_sequence_number
+                .expect("Optimistic sequence number is always set for data read from DB")
+                .try_into()
+                .unwrap(),
+            &self.metrics,
+        );
+
+        let tx_data_to_commit = extractor.to_transaction_data_to_commit().await?;
+
+        self.persist_optimistic_tx(tx_data_to_commit).await
+    }
+
+    async fn assign_optimistic_tx_global_order(
+        &self,
+        tx_digest: &TransactionDigest,
+    ) -> Result<Option<TxGlobalOrder>, IndexerError> {
+        let tx_digest_bytes = tx_digest.inner().to_vec();
+
+        let pool = self.indexer_reader.get_pool();
+
+        let mut results: Vec<TxGlobalOrder> = transactional_blocking_with_retry!(
+            &pool,
+            |conn| {
+                sql_query(
+                    r#"
+                        INSERT INTO tx_global_order (tx_digest, global_sequence_number)
+                        SELECT $1, MAX(tx_sequence_number) FROM tx_digests
+                        ON CONFLICT (tx_digest) DO NOTHING
+                        RETURNING *;
+                    "#,
+                )
+                .bind::<sql_types::Bytea, _>(&tx_digest_bytes)
+                .load::<TxGlobalOrder>(conn)
```

Comment:
You can take `Option<TxGlobalOrder>` directly

```rust
                .get_result::<TxGlobalOrder>(conn)
                .optional()
```

Requires `use diesel::OptionalExtension`.